Stay reasonably neutral in problem statement (#25)
guybedford/proposal-is-usv-string: README.md

In WebIDL, potentially ill-formed Strings are referred to using the [DOMString](https://webidl.spec.whatwg.org/#idl-DOMString) type. Well-formed strings are referred to using the [USVString](https://webidl.spec.whatwg.org/#idl-USVString) type. Due to their asymmetric value spaces, conversions from `DOMString` to `USVString` are lossy, whereas `DOMString` can represent any `USVString`. Upon conversion from `DOMString` to `USVString`, common options are to replace unpaired surrogates with the replacement character (`U+FFFD`) or to throw an error. In mixed systems, [WTF-8](https://simonsapin.github.io/wtf-8/) is often utilized to mitigate issues from too early sanitization.

> The authors of the WebAssembly [Component Model](https://github.com/WebAssembly/component-model) have, as part of the WebAssembly CG's decision-making process to advance the Component Model [to Phase 1](https://github.com/WebAssembly/proposals), achieved consensus under notable opposition (including a [formal objection](https://www.assemblyscript.org/standards-objections.html#component-model-2022-09)) to only support `USVString` on WebAssembly component boundaries. The authors of the Component Model consider Unicode text to be the best design for their users, whereas the concern expressed is that support for Java-like languages and safely interfacing with JavaScript and Web APIs is not covered by the Component Model. The authors of the Component Model have chosen the replacement strategy, so interfacing `DOMString`s with WebAssembly components, including their envisioned ESM-integration, will be subject to silent mutation of string data. Independently, there is a significant number of data encodings, network interfaces, filesystem interfaces, compile-to-JS programming languages, etc. designed for Unicode text (`USVString`) rather than `DOMString`. Therefore, interfacing JavaScript strings with such APIs is, also given the preferences of the WebAssembly CG, an increasingly common use case that suffers from the outlined conversion burdens. To detect respectively address the side-effect of silent data mutation or thrown errors early, there is a regular need for manual string validation and sanitization both within the platform and for certain userland use case scenarios. This proposal addressses both detection and sanitization.

Reviewer: Why do the minutiae of the wasm descision-making process matter here? It’s clear that you object to this decision in wasm but readers here shouldn’t need to know or care that anyone objected.

Author: I think the context is important here, because, since it is directly connected, it increases the surface area this proposal has to cover, and in part motivates the timing to pursue the proposal at this point in time. In particular, if the Component Model would revise its decision and allow passing `DOMString`s (that are also Java, Dart, Kotlin, C#, etc. strings) over WebAssembly boundaries (i.e. to each other and JS / Web APIS) without eagerly mutating them, the decision would impact the scope of the problem statement as per the problem statement. One critical aspect is, as I think, that the Component Model is on track to be used for ESM integration (hence mentioning), which would introduce the complications this proposal addresses for the entire platform when JavaScript modules are transparently replaced with WebAssembly modules, either directly or in dependencies. Do you have ideas how to integrate these relevant aspects in a form that you'd be OK with?

Reviewer: The usefulness of this proposal exists even if wasm wasn’t part of the discussion.

I think documenting wasm’s decision is sufficient, and zero context about the decision besides *maybe* a link is both sufficient and desirable.

Author: While I would agree that the proposal can be considered useful independently, all of timing, reasoning and extent of performance considerations in browser implementations are directly connected to the new situation introduced by the Component Model, as it significantly adds to scope and urgency. To illustrate, I contend that there is an alternative time line where userland implementations would be considered sufficient, say if the WebAssembly CG had more thoroughly addressed the far-reaching concerns and decided differently. The extraordinary platform complications, and why they will exist, certainly deserve a mention in my opinion. Open to restructure if you have ideas how to integrate these important aspects in a form that is more desirable to you.

Reviewer: I think that merely stating the wasm semantics with zero of the history is sufficient to convey that.

Author: I certainly recognize that there is a tension here, that, as it seems, revolves around a preference that little to zero context is provided. On the one hand, I would argue that a felt sense for zero context has little weight, and on the other that similar lack of context has led to the Component Model becoming unnecessarily suboptimal for many popular programming languages and, here in particular, JS / Web API interop affecting ECMAScript as well, now extending the complications to the entire platform. From what I've witnessed, bad outcomes are best prevented by providing related context, better more than too little, so interested parties are well-informed before decisions are made. And in this case, the context helps to answer important questions: "Why even?" and "Why now?"

Reviewer: The issue is that the "why" of WASM's decision *does not matter here*. TC39 proposal readmes aren't a pulpit for WASM opinions :-)

WASM decided. Therefore, that's the world we live in, whether it's better or worse is irrelevant. The extra (quite editorial content, because "neutral" is an editorial decision also) is just distracting to readers here.

Author: Sadly, it appears to me that we are not covering ground despite my best attempts to outline the, as I believe legitimate, reasoning for thoughtful balance. I also take slight offense by the accusation to contribute to distraction while simultaneously being subjected to distraction from the contents of my comments. Perhaps it makes sense to take a step back at this point, and see if others have opinions that would help to unstuck the arguably unproductive exchange.

In WebIDL, possibly-ill-formed Strings are referred to using the [DOMString](https://webidl.spec.whatwg.org/#idl-DOMString) type. But for interfaces which operate on Unicode text, WebIDL also defines the [USVString](https://webidl.spec.whatwg.org/#idl-USVString) type as the set of all possible sequences of Unicode Scalar Values, which are all of the Unicode code points apart from the surrogate code points (`U+0000..U+D7FF` and `U+E000..U+10FFFF`), representing Unicode text.

> The WebAssembly [Component Model](https://github.com/WebAssembly/component-model) has chosen to only support well-formed strings, even though it could also support `DOMString`. Unlike ECMAScript (JavaScript) and similar Java-like languages, some other programming languages, many data encodings, network interfaces, filesystem interfaces, etc. support only well-formed strings. Interfacing JavaScript strings with such APIs is a common use case that therefore suffers from conversion burdens. In particular, because conversion from `DOMString` to `USVString` is lossy (common options are to replace unpaired surrogates or to throw an error), there is a regular need for string validation both within the platform and for certain userland use case scenarios.

Author: What I ultimately would like to avoid is the use of language that one way or another indicates that the decisions in the Wasm CG are either good or bad. Ideally, just the facts should be stated, as neutral as possible. Perhaps:

> The authors of the WebAssembly [Component Model](https://github.com/WebAssembly/component-model) have, as part of the WebAssembly CG's decision-making process, established support for only Unicode text (`USVString`) on WebAssembly component boundaries, whereas JavaScript strings (`DOMstring`) are not supported.

For the same reason, I am also not too comfortable with terms like "many" and would rather try to avoid these. For instance, while one can casually say that there are "many" languages representing their strings as Unicode text, there are equally "many" that don't. Without quantifying in detail, the term doesn't really add anything but imply that there are more of the one than of the other. Will continue to think about a more neutral formulation, but perhaps the points I mentioned are already in your sense as well?

> The WebAssembly [Component Model](https://github.com/WebAssembly/component-model) has chosen to only support well-formed strings, even though it could also support `DOMString`. Unlike ECMAScript (JavaScript) and similar Java-like languages, some other programming languages, many data encodings, network interfaces, filesystem interfaces, etc. support only well-formed strings. Interfacing JavaScript strings with such APIs is a common use case that therefore suffers from conversion burdens. In particular, because conversion from `DOMString` to `USVString` is lossy (common options are to replace unpaired surrogates or to throw an error), there is a regular need for string validation both within the platform and for certain userland use case scenarios.

Reviewer 2: > Ideally, just the facts should be stated, as neutral as possible. Perhaps:

Sure, that suggestion seems fine.

> For the same reason, I am also not too comfortable with terms like "many"

Terms like "many" help the committee understand the extent of the issue. We do not typically solve problems that are not experienced by "many" developers. In fact, in the most recent presentation to committee, a delegate was supportive of this proposal because "there's copies of this function all over the place", indicating there are a variety of use cases that require it.

> imply that there are more of the one than of the other.

The term "many" here does not make such an implication, and I'm sure that the committee is well aware that the vast majority of web APIs consume and produce DOMStrings, so that doesn't seem worth calling out either.

> The WebAssembly [Component Model](https://github.com/WebAssembly/component-model) has chosen to only support well-formed strings, even though it could also support `DOMString`. Unlike ECMAScript (JavaScript) and similar Java-like languages, some other programming languages, many data encodings, network interfaces, filesystem interfaces, etc. support only well-formed strings. Interfacing JavaScript strings with such APIs is a common use case that therefore suffers from conversion burdens. In particular, because conversion from `DOMString` to `USVString` is lossy (common options are to replace unpaired surrogates or to throw an error), there is a regular need for string validation both within the platform and for certain userland use case scenarios.

Author: I've rephrased more of the paragraph, trying to address the different opinions. PTAL :)

> The WebAssembly [Component Model](https://github.com/WebAssembly/component-model) has chosen to only support well-formed strings, even though it could have chosen to support `DOMString` as well. Unlike ECMAScript and similar Java-like languages, some other programming languages, many data encodings, network interfaces, filesystem interfaces, etc. support only well-formed strings. Interfacing JavaScript strings with such APIs is a common use case that therefore suffers from conversion burdens. In particular because conversion from `DOMString` to `USVString` is lossy (common options are to replace unpaired surrogates or to throw an error) there is a regular need for string validation both within the platform and for certain userland use case scenarios.

Reviewer 3: Simplification
```suggestion
The WebAssembly [Component Model](https://github.com/WebAssembly/component-model) has chosen to only support well-formed strings, even though it could also support `DOMString`. Unlike ECMAScript and similar Java-like languages, some other programming languages, many data encodings, network interfaces, filesystem interfaces, etc. support only well-formed strings. Interfacing JavaScript strings with such APIs is a common use case that therefore suffers from conversion burdens. In particular, because conversion from `DOMString` to `USVString` is lossy (common options are to replace unpaired surrogates or to throw an error) there is a regular need for string validation both within the platform and for certain userland use case scenarios.
```

### Are consumers going to do anything other than convert when they encounter ill-formed strings? If so, why not provide only a conversion method with a fast path for well-formed strings?

Consumers may want to throw/error when encountering ill-formed strings. Also, consumers may want to defer the conversion or the error until later when the String is actually interpreted as Unicode text. These use cases justify the test-only method.

> ### How does this proposal relate to the WebAssembly Component Model?

Reviewer: This PR seems fine to me now, if this entire superfluous section is removed.

Author: There are several direct connections to this proposal in this section of the FAQ (that I moved there from the problem statement to do ya'll a favor btw). In addition to explaining why the CM is probably the most notable precedent for this proposal, why timings overlap, why this is special, why performance is important, etc., the paragraph also aids future discussion. Today, this proposal uses the CM as an argument, and it is likely that tomorrow the CM will use this proposal as an argument to bolster its choices because "but you can check". This is highly relevant context that, if withheld for editorial reasons, will not be obvious and likely lead to worse results overall than if it was provided. Please let's be responsible.

Reviewer: As a JS developer, I should not need to understand or read anything about wasm beyond "wasm requires well-formed strings". None of the info about the component model is relevant to me. A TC39 proposal is not an appropriate place to create leverage for an argument within wasm.

Author: This is not intended as leverage, really. If you can point me to sentences that feel like it to you, please tell me, and I'll do my best to phrase as objectively as I humanly can. Hardly anyone knows about the nitty gritty details in play here (this proposal helps to some extent), even less people know how all this is connected, but exactly this knowledge is important, now more than ever with the CM doing things differently than the platform has done it before. As a JS developer, you should care because you have dependencies, and these dependencies have dependencies, you upgrade them, you bundle them. This string stuff is subtle, but also fundamental, and errors are delayed, almost impossible to spot. With the CM, JS developers need a bunch of highly specialized knowledge now that they didn't necessarily need to care about before. I do think that we are acting responsible when we try our best to let them know about the what and why, so they can address it, either in code or in the spec. Please give me a realistic chance to improve this FAQ section. I think it's extremely important that people are aware.

Reviewer: I think the place that awareness should be spread is a wasm arena, not a JS one.

Author: Please, this is about helping JS devs. To help TC39. ECMAScript. I am out of options in Wasm, nobody cares there :(

Reviewer: That's exactly the point - if wasm's decisions here aren't going to change, why is this extra context useful to JS devs? All they need to know is _what_ a well-formed string is, how to check for it, and how to sanitize one if needed - which this proposal already provides for.

Author: Given that the CM is in phase 1, it's too early to know for sure whether Wasm's decisions are going to last. And as I said in my prior comment, JS devs, in particular those intending to consume Wasm modules, even more so those with an interest to use JS and Wasm in tandem (which is a fantastic use case btw and one of Wasm's stated goals), should care:

> As a JS developer, you should care because you have dependencies, and these dependencies have dependencies, you upgrade them, you bundle them. This string stuff is subtle, but also fundamental, and errors are delayed, almost impossible to spot. With the CM, JS developers need a bunch of highly specialized knowledge now that they didn't necessarily need to care about before. I do think that we are acting responsibly when we try our best to let them know about the what and why, so they can address it, either in code or in the spec.

If they are kept in the dark, this will just remain the self-fulfilling prophecy it already is: Wasm does its thing, while all attempts to inform or include JS devs are stonewalled. Greetings to the same Wasm folks up- and downvoting here btw. :)

Reviewer: If wasm makes decisions that make JS interop harder *in practice*, that hurts wasm devs, not JS devs, because that will just inhibit use of wasm. I don't see what JS devs can do about it - nor do I think that the entry to the wasm funnel will be on this particular TC39 proposal.

Author: That doesn't sound like a good outcome, rather like an argument to better have an FAQ entry so nobody is hurt unknowingly. And for sure this proposal is the entry point: Want to know what the problem is? Direct here. Ran into an issue? Direct here. Component Model? Soon directs here. In a sense, JS and Wasm meet exactly here, where the problem worked around.
